### PR TITLE
 🌱 E2E: Decouple control plane status check logic

### LINKED ISSUE #4749: Decouple control plane status check logic
**User Story**

In e2e test framework's `ApplyClusterTemplateAndWait`, the logic for control plane is tightly coupled with `kubeadmControlPlane` implementation. This is not suitable when we want to test clusters where the control plane is not kubeadm. Eg. In capz, for [testing AKS clusters](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/611) where the control plane is an `AzureManagedControlPlane`. In CAPA this is fixed by not using `ApplyClusterTemplateAndWait` and calling the individual functions like `DiscoveryAndWaitForCluster`. But, this will result in duplicate code. 

**Detailed Description**

Decouple control plane status check logic, and provide the ability to inject custom checks for control plane if needed.  Default to `KubeadmControlPlane` if none provided.

**Anything else you would like to add:**

/kind feature

### PR DESCRIPTION
**Pull request by shysank**
**What this PR does / why we need it**:

In e2e test framework's `ApplyClusterTemplateAndWait`, the logic for control plane is tightly coupled with kubeadmControlPlane implementation. This is not suitable when we want to test clusters where the control plane is not kubeadm. Eg. In capz, for [testing AKS clusters](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/611) where the control plane is an `AzureManagedControlPlane`. In CAPA this is fixed by not using `ApplyClusterTemplateAndWait` and calling the individual functions like `DiscoveryAndWaitForCluster`. But, this will result in duplicate code. This PR tries to address this concern by decoupling control plane status check logic, and providing the ability to inject custom checks for control plane if needed.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4749  ~There is no issue created to track this one, happy to create on if needed.~

### CODE REVIEW

**Comment by shysank** [comment]
cc @nader-ziada @CecileRobertMichon

File: test/framework/clusterctl/clusterctl_helpers.go [hunk 1]
```diff
@@ -272,3 +276,25 @@ func ApplyClusterTemplateAndWait(ctx context.Context, input ApplyClusterTemplate
 		Cluster: result.Cluster,
 	}, input.WaitForMachineDeployments...)
 }
+
+// setDefaults sets the default values for ApplyClusterTemplateAndWaitInput if not set.
```
Comment (fabriziopandini) [inline test/framework/clusterctl/clusterctl_helpers.go:280]: nit: Just add a comment reporting that the current implementation of the default waiters are designed for a KCP control plane.

**Review by fabriziopandini — COMMENTED** [review]
@shysank thanks for this PR!
I'm ok with the current change, but I would like to have an issue tracking this problem and asking for having default waiters based on the ClusterAPI contract only. Do you mind open one?
/lgtm

**Pull request merged**